### PR TITLE
Batched linear system of equations solver (torch.bgesv)

### PR DESCRIPTION
**Pull request by zou3519**
Implements torch.bgesv, a batched linear system of equations solver.

Adds bindings for MAGMA's gesv_batched function for CUDA.
For CPU, runs THLapack(gesv) in a for loop.

I decided to not build this into `torch.gesv` but if we want to I can change that.

cc @apaszke 

### Test Plan
New unit tests:
- CPU test against `torch.gesv` and specifying outputs
- GPU test for the same thing
- autograd test for derivative

### CODE REVIEW

**Comment by apaszke** [comment]
cc @fritzo

**Comment by fritzo** [comment]
cc @tbrx @dwd31415

File: aten/src/TH/generic/THTensorLapack.c [hunk 1]
```diff
@@ -101,6 +101,34 @@ static THTensor *THTensor_(cloneColumnMajor)(THTensor *self, THTensor *src)
...
+    THTensor_(retain)(self);
+  }
+
+  int64_t size[3] = { src->size[0], src->size[1], src->size[2] };
+  int64_t stride[3] = { matrix_size, 1, src->size[1] };
+
+  THTensor_(resizeNd)(self, 3, size, stride);
+  THTensor_(copy)(self, src);
+  return self;
+}
```
Comment (apaszke) [inline aten/src/TH/generic/THTensorLapack.c]: I think it'd be better to rearrange some statements in this function. I'd move the `size` and `stride` arrays to the top and used that both to check if `self` matches what we expect, and in a call to `resizeNd` (less duplication, fewer chances for errors).

File: test/test_torch.py [hunk 2]
```diff
@@ -2092,6 +2092,37 @@ def test_gesv(self):
...
         self.assertEqual(res1, tb)
 
+    def _test_bgesv(self, cast):
+        # test against gesv
+        A = cast(torch.randn(1, 5, 5))
+        b = cast(torch.randn(1, 5, 10))
+        x_exp, LU_exp = torch.gesv(b.squeeze(0), A.squeeze(0))
+        x, LU = torch.bgesv(b, A)
+        self.assertEqual(x, x_exp.unsqueeze(0))
+        self.assertEqual(LU, LU_exp.unsqueeze(0))
```
Comment (apaszke) [inline test/test_torch.py]: Can you make a test that compares to `gesv` in a short loop (let's say we have a 4 elem batch)?

**Review by apaszke — COMMENTED** [review]
That looks great, thanks for wrapping this up so quickly! Only a few minor comments.

File: aten/src/THC/generic/THCTensorMathMagma.cu [hunk 3]
```diff
@@ -98,6 +121,76 @@ THC_API void THCTensor_(gesv)(THCState *state, THCTensor *rb_, THCTensor *ra_, T
...
+  int **ipiv_array = th_magma_malloc_pinned<int *>(batch_count);
+
+  for (int64_t i = 0; i < batch_count; i++) {
+    a_array[i] = &a_data[i * n * n];
+    b_array[i] = &b_data[i * n * nrhs]; 
+    ipiv_array[i] = &ipiv_data[i * n];
+  }
+
+  magma_queue_t magma_queue;
+  magma_queue_create_from_cuda(
```
[the remark this answers is not in the document]
Comment (zou3519) [reply aten/src/THC/generic/THCTensorMathMagma.cu]: Not yet. Will measure and report back.

[the remark this answers is not in the document]
Comment (zou3519) [reply aten/src/THC/generic/THCTensorMathMagma.cu]: The overhead is ~350000 microseconds the first time it's called (on my machine), and ~6 microseconds every time after that. Is that reasonable?

**Comment by colesbury** [comment]
A few high-level comments:

The equivalent NumPy function is [np.linalg.solve](https://docs.scipy.org/doc/numpy-1.13.0/reference/generated/numpy.linalg.solve.html)

1) We should prefer one function that handles batched and non-batched computation.
2) We should treat the left-most dimensions as part of the batch. For example, the 4-d tensor `A x B x M x M` should be treated as batch-size `(A x B)`.
3) It would be good to support broadcasting
4) When adding new functions, consider adding them as native functions to ATen.

You don't have to block the PR on these suggestions, but if it's not too much extra work see if you can merge gesv and bgesv so that we don't unnecessarily expand the public API.

**Comment by zou3519** [comment]
Closing this in favor of #4612

**Pull request closed without merging**